src/Shared/CopyOnWriteDictionary.cs

```diff
@@ -200,53 +179,37 @@ internal bool IsDummy
         /// <summary>
         /// Comparer used for keys
         /// </summary>
-        internal IEqualityComparer<K> Comparer { get; private set; }
+        internal IEqualityComparer<K> Comparer
+        {
+            get => ReadOperation.KeyComparer;
+            private set => backing = WriteOperation.WithComparers(keyComparer: value, valueComparer: backing.ValueComparer);
+        }
 
         /// <summary>
         /// Gets the backing dictionary for reading.
         /// </summary>
-        private CopyOnWriteBackingDictionary<K, V> ReadOperation
+        private ImmutableDictionary<K, V> ReadOperation
         {
             get
             {
                 ErrorUtilities.VerifyThrow(!IsDummy || backing == null || backing.Count == 0, "count"); // check count without recursion
-#if DEBUG
-                if (s_forceWrite)
-                {
-                    if (!IsDummy)
-                    {
-                        return WriteOperation;
-                    }
-                }
-#endif
-                if (backing == null)
-                {
-                    return CopyOnWriteBackingDictionary<K, V>.ReadOnlyEmptyInstance;
-                }
 
-                return backing;
+                return backing ?? ImmutableDictionary<K,V>.Empty;
             }
         }
 
         /// <summary>
         /// Gets the backing dictionary for writing.
         /// </summary>
-        private CopyOnWriteBackingDictionary<K, V> WriteOperation
+        private ImmutableDictionary<K, V> WriteOperation
         {
             get
             {
                 ErrorUtilities.VerifyThrow(!IsDummy, "dummy");
 
                 if (backing == null)
                 {
-                    backing = new CopyOnWriteBackingDictionary<K, V>(capacity, Comparer);
-                }
-                else
-                {
-                    lock (((ICollection)backing).SyncRoot)
-                    {
-                        backing = backing.CloneForWriteIfNecessary();
-                    }
+                    backing = ImmutableDictionary.Create<K,V>(Comparer);
```

Comment:
Does this keep all the mappings?

Reply:
This code was unreachable due to `backing` being non-nullable. The entire `WriteOperation` has been deleted.